lib/dry/core/memoizable.rb

```diff
@@ -43,11 +43,12 @@ def define_memoizable_names!
 
             if meth.parameters.size > 0
               define_method(name) do |*args|
-                __memoized__[:"#{name}_#{args.hash}"] ||= super(*args)
+                key = :"#{name}_#{args.hash}"
+                __memoized__.fetch(key) { __memoized__[key] = super(*args) }
               end
             else
               define_method(name) do
-                __memoized__[name] ||= super()
+                __memoized__.fetch(name) { __memoized__[name] = super() }
```

Comment:
this allocates a proc, it'd be better to do `__memoized__.key?(name) ? __memoized__[name] : super()`

Reply:
@solnic This doesn't look to me like it allocates a proc:

```ruby
[1] pry(main)> hsh = {}
=> {}
[2] pry(main)> hsh.fetch(:solnic) { hsh[:solnic] = "hello mate" }
=> "hello mate"
[3] pry(main)> hsh
=> {:solnic=>"hello mate"}
[4] pry(main)> hsh.fetch(:solnic)
=> "hello mate"
```

Reply:
@solnic or do you mean that a proc gets created in memory? If so, then yeah, you're right. I'll make it simpler.

Reply:
@solnic this is actually make no difference or my benchmarking sucks

```ruby
require "benchmark/ips"

hash = { 1 => true }

Benchmark.ips do |x|
  x.report("fast code description") { hash.fetch(1) }
  x.report("slow code description") { hash.fetch(1) { 2 } }
  x.compare!
end

__END__

Calculating -------------------------------------
fast code description
                          7.891M (± 7.4%) i/s -     39.315M in   5.011015s
slow code description
                          8.029M (± 7.0%) i/s -     40.046M in   5.014197s

Comparison:
slow code description:  8028735.8 i/s
fast code description:  7890573.3 i/s - same-ish: difference falls within error
```

Reply:
@flash-gordon I think solnic was talking more about memory usage than speed?

Reply:
@timriley it seems like a lesser issue to me :)

Reply:
yes, mostly about memory, I know there's small perf diff, but since I want this in rom so we can remove rom/support/memoizable, I'm just being extra careful with additional object allocations. I've been obsessing with this and we clearly see benefits (rom takes WAY less memory than AR)